// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: tragedylooper/v1/incident.proto

package v1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on Incident with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Incident) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Incident with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in IncidentMultiError, or nil
// if none found.
func (m *Incident) ValidateAll() error {
	return m.validate(true)
}

func (m *Incident) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetConfig()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, IncidentValidationError{
					field:  "Config",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, IncidentValidationError{
					field:  "Config",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetConfig()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return IncidentValidationError{
				field:  "Config",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for HasTriggeredThisLoop

	if len(errors) > 0 {
		return IncidentMultiError(errors)
	}

	return nil
}

// IncidentMultiError is an error wrapping multiple validation errors returned
// by Incident.ValidateAll() if the designated constraints aren't met.
type IncidentMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m IncidentMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m IncidentMultiError) AllErrors() []error { return m }

// IncidentValidationError is the validation error returned by
// Incident.Validate if the designated constraints aren't met.
type IncidentValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e IncidentValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e IncidentValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e IncidentValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e IncidentValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e IncidentValidationError) ErrorName() string { return "IncidentValidationError" }

// Error satisfies the builtin error interface
func (e IncidentValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sIncident.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = IncidentValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = IncidentValidationError{}

// Validate checks the field values on IncidentConfig with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *IncidentConfig) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on IncidentConfig with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in IncidentConfigMultiError,
// or nil if none found.
func (m *IncidentConfig) ValidateAll() error {
	return m.validate(true)
}

func (m *IncidentConfig) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Name

	// no validation rules for Description

	// no validation rules for Day

	if all {
		switch v := interface{}(m.GetCondition()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, IncidentConfigValidationError{
					field:  "Condition",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, IncidentConfigValidationError{
					field:  "Condition",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetCondition()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return IncidentConfigValidationError{
				field:  "Condition",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetEffect()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, IncidentConfigValidationError{
					field:  "Effect",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, IncidentConfigValidationError{
					field:  "Effect",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetEffect()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return IncidentConfigValidationError{
				field:  "Effect",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for IsMandatory

	// no validation rules for IsMainPlotIncident

	// no validation rules for IsSpecialRule

	if len(errors) > 0 {
		return IncidentConfigMultiError(errors)
	}

	return nil
}

// IncidentConfigMultiError is an error wrapping multiple validation errors
// returned by IncidentConfig.ValidateAll() if the designated constraints
// aren't met.
type IncidentConfigMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m IncidentConfigMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m IncidentConfigMultiError) AllErrors() []error { return m }

// IncidentConfigValidationError is the validation error returned by
// IncidentConfig.Validate if the designated constraints aren't met.
type IncidentConfigValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e IncidentConfigValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e IncidentConfigValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e IncidentConfigValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e IncidentConfigValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e IncidentConfigValidationError) ErrorName() string { return "IncidentConfigValidationError" }

// Error satisfies the builtin error interface
func (e IncidentConfigValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sIncidentConfig.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = IncidentConfigValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = IncidentConfigValidationError{}
